### ISSUE TITLE
Make experiments runnable without firing a live submission

### ISSUE THREAD

**Issue by vltnkiz** [vltnkiz/numerai#25]
Part of #24

## Question

How do we run zemir experiments **without firing a live Numerai submission**, and where do they run?

Two hard blockers sit under every quantitative ticket on this map:

1. **`scripts/run_pipeline.py` always submits.** There is no dry-run flag — `run_pipeline()` calls `_submit_predictions()` unconditionally at `pipeline.py:180`. Any comparison run today would upload to slot `zemir_01` and overwrite the round's real submission.
2. **Nothing runs locally.** System Python has no `pandas`/`numpy`/`pyarrow`; there is no venv, conda, pyenv, or uv environment on the machine. The only place this code has ever executed is GitHub Actions.

Resolve by deciding and doing:

- A no-submit path — a `--no-submit`/`--dry-run` flag, an injected null submitter, or splitting `run_pipeline` so submission is a separate call. Which one, and does it change the shape of `PipelineResult`?
- Where experiments execute: local venv (needs ~6.2 GB of parquet on disk, one-time download) vs a separate manually-dispatched CI workflow. Local iterates faster; CI is where the code demonstrably works.
- If local: create the environment and confirm `pip install -e ./prod/zemir_01` then a full no-submit run succeeds.

**Answer must record**: the mechanism chosen, the exact command to run an experiment safely, and where datasets live.

**Comment by vltnkiz** [comment]
## Resolution

Experiments run **locally**, through a **second entrypoint that has no submission code in it**. Landed in 1bfc36f on `main`.

### Mechanism: structural split, not a flag

`run_pipeline` no longer submits. It trains, scores, predicts, neutralizes, rank-normalizes and writes artifacts, then returns. `submit_predictions` (was `_submit_predictions`) is the only function in the package that uploads, and the only caller is `scripts/run_pipeline.py` — the entrypoint the scheduled workflow invokes.

A `--dry-run` flag was rejected: it leaves submission as the *default* and makes safety depend on remembering. With the split, an experiment cannot fire a live submission however it is invoked, because there is no code path from it to an upload.

**`PipelineResult` did change shape**: `submission: SubmissionResult` is gone (the caller composes the two steps), and `run_dir: Path` is added so the caller knows where to write `submission.json`.

The `MIN_VALIDATION_MEAN_CORR` gate also moved out of the pipeline into `scripts/run_pipeline.py`, between the run and the submit. The gate exists to stop a bad model *being submitted*; now that the pipeline cannot submit, it had nothing to stop, and keeping it inside would have forced a third config profile purely to disable it for experiments.

### Configuration: named profiles, not parameters

Settings moved out of the entrypoints into one flat frozen `PipelineConfig` in `zemir/config.py`, with named instances `LIVE` and `SMOKE`. `run_pipeline(config, *, run_id, trainers, runs_dir, neutralizers)`.

This was a deliberate reversal of an earlier draft that threaded `max_eras` and `**xgboost_overrides` through as parameters. Every remaining ticket on this map adds another dimension to vary, and this map's destination is *measured comparison* — that needs a configuration you can name, pass around, serialize and diff, not a 15-parameter function.

Constraints agreed while deciding this, which matter because #24's history includes a `RunConfig` that was deleted in 40df371 for being too heavy:
- **Exactly one dataclass, flat.** Scalars and one dict. No sub-configs, ever — the old one was four nested dataclasses, and that nesting is what sank it. If it ever needs nesting, that is the signal it has grown wrong.
- **No `run_id` inside it.** A config names a *configuration*; `run_id` names an *instance*. Conflating them was a flaw in the deleted design.
- **`SUBMISSION_MODEL_SLOT` and `MIN_VALIDATION_MEAN_CORR` stay out**, as module constants read only by the live script. A pipeline config carrying a submission target would mean every `SMOKE` instance holds a live slot name it must never use.
- **`neutralizers` stays a `run_pipeline` parameter**, not a config field — deciding it must be a named feature set would pre-constrain #29, which might want something no named set expresses (e.g. "the 20 highest-exposure features").

### Exact commands

```bash
# one-time
python -m venv ~/ml-venv
~/ml-venv/bin/pip install -e prod/zemir_01

cd prod/zemir_01
~/ml-venv/bin/python scripts/run_experiment.py --model ensemble --smoke   # ~12s, code path only
~/ml-venv/bin/python scripts/run_experiment.py --model ensemble           # ~1h, real numbers
```

`scripts/run_pipeline.py` is the **live** command and submits on every invocation. Its name, arguments and CI invocation are unchanged, so `.github/workflows/weekly-pipeline.yml` needed no edit.

### Where things live

- **Datasets**: `datasets/<version>/` in the repo, gitignored. v5.0 is **already downloaded**, 5.9 GB. `train.parquet` 2,746,268 rows / `validation.parquet` 4,114,072 rows / `live.parquet` 6,976 rows, 2,416 columns each; the `small` feature set is 42 features. `train`/`validation` are cached; `live.parquet` re-downloads every run by design.
- **Environment**: `~/ml-venv` (Python 3.11.2) — pandas 3.0.5, numpy 2.4.6, xgboost 3.2.0, scikit-learn 1.9.0, scipy 1.17.1, pyarrow 25.0.1, numerapi 3.1.1. `pip install -e ./prod/zemir_01` succeeds.
- **Experiment artifacts**: `runs/experiments/<run_id>-<model>[-smoke]/`, apart from the production record in `runs/<run_id>/`.
- **Credentials**: `prod/zemir_01/.env`, read only by `submit_predictions`.

### Smoke scale — the finding that changed the ticket

A full era-boosted run takes **~1 hour** (40 XGBoost refits over 2.7M rows), which made the `ensemble` path effectively unverifiable during development. `SMOKE` trims to the 40 most recent eras and 2 boosting iterations, running the **identical** code path in ~12 s. Its scores are meaningless and must never be compared against a full run — it answers "does this run at all", which is what iterating on pipeline code actually needs.

Measured, 8-core machine: `linear` full **15 s** · `ensemble --smoke` **12 s** · `ensemble` full **~1 h** (estimate, not measured to completion).

### Runs are now self-describing

Every run writes `config.json` beside its scores, recording the config fields, the models fitted (`sorted(trainers)`) and the resolved neutralizer count. Nothing previously recorded what settings produced a `validation_score.json`, which would have made #28's comparisons unattributable.

### Verification

1. **Experiment path cannot submit** — `numerapi.NumerAPI.upload_predictions` and `numerapi.base_api.Api.upload_predictions` replaced with a raising stub; a full `ensemble` run completed, 6,976 live rows, no submission attempted.
2. **Live path unchanged** — `scripts/run_pipeline.py` driven with the upload stubbed: 6,976 rows, columns `['id', 'prediction']`, correct `model_id`, predictions in `(7.17e-05, 0.99993)`, validation `mean_corr=0.0128 sharpe=0.7176` — identical to before the refactor. `submission.json` still written.
3. **Relocated gate works** — with `MIN_VALIDATION_MEAN_CORR` raised to 0.99, the live script raised `ValidationScoreBelowThreshold` and did not submit.
4. **Multi-model path executed for the first time** — `_combine_predictions` had never run in production (live has only ever submitted `linear`). It works.

### Notes for later tickets

- Nothing **automatically** enforces the no-submit boundary — the guarantee is structural plus documentation. Deliberately deferred (no test infrastructure exists in this repo, and it wasn't possible to write any until this ticket made local runs work). Recorded as fog on #24.
- `run_pipeline` still neutralizes **only live predictions**; validation is scored *before* combination and *without* neutralization. That gap is exactly #28's subject, now unblocked.
- Every experiment re-downloads `live.parquet` (~8 MB, force=True). Harmless, but a validation-only harness may want to skip it.
